ambry-router/src/main/java/com/github/ambry/router/UndeleteOperation.java

```diff
@@ -141,19 +361,19 @@ private int getPrecedenceLevel(RouterErrorCode routerErrorCode) {
     switch (routerErrorCode) {
       case BlobAuthorizationFailure:
         return 0;
-      case BlobDeleted:
+      case LifeVersionConflict:
         return 1;
       case BlobExpired:
         return 2;
-      case BlobUpdateNotAllowed:
-        return 3;
       case AmbryUnavailable:
-        return 4;
+        return 3;
       case UnexpectedInternalError:
-        return 5;
+        return 4;
       case OperationTimedOut:
-        return 6;
+        return 5;
       case BlobDoesNotExist:
+        return 6;
+      case BlobNotDeleted:
```

Comment:
I feel `BlobNotDeleted` may have higher precedence level than `OperationTimedOut`

Reply:
this one is quite tricky. Two errors in one dc would render undelete operation a failure. Say if we have one `BlobNotDeleted` and one `OperationTimeOut`, it might be because there is only one host that doesn't have this blob deleted, but the other two are deleted and unfortunately one of these two returns `OperationTimeOut`. In this case, I think it's better to return `OperationTimeOut`.

Reply:
I think `BlobNotDeleted` means one server holds the blob but this blob is not deleted yet. If undelete succeeds on all servers except for one server that returns `BlobNotDeleted`, we still consider whole operation a success?

Reply:
you are right, `BlobNotDeleted` does mean that. And `BlobNotDeleted` is not a terminating error, which means if we see this error, the undelete operation will continue. If undelete operation sees more than one error on one dc (assume there are 3 replicas), then it will fail the operation. But if undelete operation only sees one `BlobNotDeleted` and two success in one dc, then it's still considered a success.